src/protocol/attribution/apply_attribution_window.rs

```diff
@@ -0,0 +1,303 @@
+use super::{
+    do_the_binary_tree_thing,
+    input::{MCApplyAttributionWindowInputRow, MCApplyAttributionWindowOutputRow},
+};
+use crate::{
+    error::Error,
+    ff::Field,
+    protocol::{
+        boolean::{
+            bitwise_greater_than_constant, random_bits_generator::RandomBitsGenerator,
+            BitDecomposition, RandomBits,
+        },
+        context::Context,
+        BasicProtocols, RecordId,
+    },
+    secret_sharing::Arithmetic,
+};
+use futures::future::try_join_all;
+use std::iter::{repeat, zip};
+
+/// This protocol applies the specified attribution window to trigger events. All trigger values of
+/// events that are outside the window will be replaced with 0, hence will not be attributed to
+/// corresponding source events in the later aggregation stages.
+///
+/// # Errors
+/// Fails if sub-protocols fails.
+#[allow(dead_code)]
+async fn apply_attribution_window<F, C, T>(
+    ctx: C,
+    input: &[MCApplyAttributionWindowInputRow<F, T>],
+    attribution_window_seconds: u32,
+) -> Result<impl Iterator<Item = MCApplyAttributionWindowOutputRow<F, T>> + '_, Error>
+where
+    F: Field,
+    C: Context + RandomBits<F, Share = T>,
+    T: Arithmetic<F> + BasicProtocols<C, F>,
+{
+    let mut t_deltas = prefix_sum_time_deltas(&ctx, input).await?;
+
+    let trigger_values =
+        zero_out_expired_trigger_values(&ctx, input, &mut t_deltas, attribution_window_seconds)
+            .await?;
+
+    Ok(input.iter().zip(trigger_values).map(|(x, value)| {
+        MCApplyAttributionWindowOutputRow::new(
+            x.is_trigger_report.clone(),
+            x.helper_bit.clone(),
+            x.breakdown_key.clone(),
+            value,
+        )
+    }))
+}
+
+/// Computes time deltas from each trigger event to its nearest matching source event.
+///
+/// # Errors
+/// Fails if the multiplication fails.
+async fn prefix_sum_time_deltas<F, C, T>(
+    ctx: &C,
+    input: &[MCApplyAttributionWindowInputRow<F, T>],
+) -> Result<Vec<T>, Error>
+where
+    F: Field,
+    C: Context,
+    T: Arithmetic<F> + BasicProtocols<C, F>,
+{
+    let num_rows = input.len();
+
+    // Pre-compute `is_trigger_bit * helper_bit'.
+    // Later in the prefix-sum loop, this vector is updated in each iteration to help
+    // accumulate timedeltas and determine when to stop accumulating.
```

Comment:
We already did this back in the accumulate step! I'll bet that (someday) we can find a way of just doing this one time to save on mults.